crates/compression/src/compress.rs

```diff
@@ -0,0 +1,95 @@
+use fuel_core_types::{
+    blockchain::block::Block,
+    fuel_compression::CompressibleBy,
+    fuel_tx::{
+        Bytes32,
+        Transaction,
+    },
+};
+
+use crate::{
+    context::{
+        compress::CompressCtx,
+        prepare::PrepareCtx,
+    },
+    eviction_policy::CacheEvictor,
+    ports::{
+        EvictorDb,
+        TemporalRegistry,
+        UtxoIdToPointer,
+    },
+    tables::{
+        PerRegistryKeyspace,
+        RegistrationsPerTable,
+    },
+    CompressedBlockPayload,
+    Header,
+};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Only the next sequential block can be compressed")]
+    NotLatest,
+    #[error("Unknown compression error")]
+    Other(#[from] anyhow::Error),
+}
+
+pub trait CompressDb: TemporalRegistry + UtxoIdToPointer {}
+impl<T> CompressDb for T where T: TemporalRegistry + UtxoIdToPointer {}
+
+pub async fn compress<D: CompressDb + EvictorDb>(
+    db: D,
+    block: &Block,
+) -> Result<Vec<u8>, Error> {
+    // if *block.header().height() != db.next_block_height()? {
+    //     return Err(Error::NotLatest);
+    // }
+
+    let target = block.transactions().to_vec();
+
+    let mut prepare_ctx = PrepareCtx {
+        db,
+        accessed_keys: PerRegistryKeyspace::default(),
+    };
+    let _ =
+        <Vec<Transaction> as CompressibleBy<_>>::compress_with(&target, &mut prepare_ctx)
+            .await?;
+
+    let mut ctx = CompressCtx {
+        db: prepare_ctx.db,
+        cache_evictor: CacheEvictor {
+            keep_keys: prepare_ctx.accessed_keys,
+        },
+        changes: Default::default(),
+    };
+    let transactions = target.compress_with(&mut ctx).await?;
+    let registrations = ctx.changes;
+    let registrations = RegistrationsPerTable::try_from(registrations)?;
+
+    // Apply changes to the db
+    registrations.write_to_registry(&mut ctx.db)?;
+
+    // Construct the actual compacted block
+    let compact = CompressedBlockPayload {
+        registrations,
+        registrations_root: Bytes32::default(), /* TODO: https://github.com/FuelLabs/fuel-core/issues/2232 */
+        header: Header {
```

Comment:
Just to clarify: do you mean `PartialBlockHeader` or `BlockHeader`? I.e. should we also include generated fields?

Reply:
Hmm, I thought about `BlockHeader`, but maybe you are right and it will be enough to use `PartialBlockHeader`. The problem with `PartialBlockHeader` is that it is not versioned, but I think maybe it is okay since we have version in the compressed payload=)

Reply:
Changed to use `PartialBlockHeader` in e036b824b9